File: src/ultraloom/journal.py
"""The run journal: one JSONL line per node.

Deliberately one thing instead of two — the same file is the log you read to
evaluate a run and the only source a resume reads from.
"""

from __future__ import annotations

import dataclasses
import hashlib
import json
from collections.abc import Mapping
from dataclasses import dataclass
from pathlib import Path


class JournalError(ValueError):
    """Raised for a journal file that cannot be read or written."""


def _unserializable(value: object) -> object:
    """Refuse a value JSON cannot express, instead of inventing one for it.

    The tempting fallback is `repr`, but a default `repr` carries the object's
    memory address: the same payload would hash differently in a new process, so
    a resume would silently redo finished work — and a collision would silently
    skip a node. Both are invisible, so the journal says so at write time.
    """
    raise JournalError(f"cannot serialize a value of type {type(value).__name__}")


@dataclass(frozen=True, slots=True)
class Entry:
    """What one node did."""

    node: str
    kind: str
    input_hash: str
    delta: Mapping[str, object]
    outcome: str
    tools: str | None
    effort: str | None
    tokens: int
    seconds: float
    detail: str | None


def input_hash(node: str, data: object) -> str:
    """A stable fingerprint of the input a node saw.

    Keys are sorted, so a hash never depends on field ordering — a resume that
    turned on dict order would replay the wrong node without saying so. Python's
    own `hash()` is salted per process and deliberately not used here.

    Raises:
        JournalError: if the payload holds a value JSON cannot express.
    """
    payload: object = data
    if dataclasses.is_dataclass(data) and not isinstance(data, type):
        payload = dataclasses.asdict(data)
    blob = json.dumps({"node": node, "data": payload}, sort_keys=True, default=_unserializable)
    return hashlib.sha256(blob.encode("utf-8")).hexdigest()


@dataclass(frozen=True, slots=True)
class Journal:
    """Append-only JSONL, read back whole."""

    path: Path

    def append(self, entry: Entry) -> None:
        """Add one line. Creates the file and its parents on first write.

        Raises:
            JournalError: if the entry holds a value JSON cannot express. It is
                refused before the file is touched, so a bad entry leaves no
                half-written line behind.
        """
        line = json.dumps(dataclasses.asdict(entry), sort_keys=True, default=_unserializable)
        self.path.parent.mkdir(parents=True, exist_ok=True)
        # LF regardless of platform: the journal is a data format whose bytes a
        # resume and the golden-journal test compare, not a text document.
        with self.path.open("a", encoding="utf-8", newline="\n") as handle:
            handle.write(line + "\n")

    def entries(self) -> tuple[Entry, ...]:
        """Every line, in order. An absent file reads as empty."""
        if not self.path.exists():
            return ()
        found: list[Entry] = []
        text = self.path.read_text(encoding="utf-8", newline="\n")
        for number, line in enumerate(text.splitlines(), start=1):
            if not line.strip():
                continue
            try:
                found.append(Entry(**json.loads(line)))
            except (json.JSONDecodeError, TypeError) as error:
                message = f"{self.path}: line {number} is not a journal entry"
                raise JournalError(message) from error
        return tuple(found)

    def lookup(self, node: str, node_input_hash: str) -> Entry | None:
        """The most recent entry for this node and this input, if any."""
        for entry in reversed(self.entries()):
            if entry.node == node and entry.input_hash == node_input_hash:
                return entry
        return None

File: internal/journal/journal.go
// Package journal reads the run journal: one JSONL line per node.
//
// The reading half only. `session-start` announces what is waiting and writes
// nothing, and the writer -- with the input hashing a resume depends on --
// belongs to the stages that walk a run.
package journal

import (
	"encoding/json"
	"fmt"
	"os"
	"slices"
	"strings"
)

// Entry is journal.py's Entry, field for field, in the JSON spelling on disk.
//
// Tools, Effort and Detail are pointers because the Python side writes
// `str | None` there and the difference carries a decision: `Pending` reads a
// nil Detail as "this pause asked nothing", which is not the same as an empty
// question.
type Entry struct {
	Node      string         `json:"node"`
	Kind      string         `json:"kind"`
	InputHash string         `json:"input_hash"`
	Delta     map[string]any `json:"delta"`
	Outcome   string         `json:"outcome"`
	Tools     *string        `json:"tools"`
	Effort    *string        `json:"effort"`
	Tokens    int            `json:"tokens"`
	Seconds   float64        `json:"seconds"`
	Detail    *string        `json:"detail"`
}

// Entries is journal.py's `entries`: every line, in order.
//
// An absent file is empty and not an error, and a line that cannot be decoded
// ends the read with its number named. Both are that module's decisions: a run
// that never started has no journal, and one unreadable line makes the whole
// journal unreadable rather than being skipped -- a journal read past its
// damage would answer about a run nobody can reconstruct. The caller decides
// what that costs; `session_start.py` prints the error and carries on with the
// other runs, so one damaged file hides its own lines and no others.
//
// A well-formed object whose keys do not match `Entry` is damage too, the same
// way `Entry(**json.loads(line))` raises TypeError for a missing or an
// unexpected keyword. It is stricter about types on top of that -- Python never
// checks its annotations and takes a string for `tokens`, this does not.
//
// The whole file is read at once, as `read_text` does, and no line length is
// refused: nothing caps what `append` writes, since `delta` is a node's own
// output.
func Entries(path string) ([]Entry, error) {
	text, err := os.ReadFile(path)
	if err != nil {
		if os.IsNotExist(err) {
			return nil, nil
		}
		return nil, fmt.Errorf("reading %s: %w", path, err)
	}

	var found []Entry
	// A blank line is skipped but still counted, so the number in an error is
	// the line a reader finds in the file.
	for number, line := range strings.Split(string(text), "\n") {
		if strings.TrimSpace(line) == "" {
			continue
		}
		entry, err := decode(line)
		if err != nil {
			return nil, fmt.Errorf("%s: line %d is not a journal entry: %w", path, number+1, err)
		}
		found = append(found, entry)
	}
	return found, nil
}

// entryKeys are the ten keys `Entry` is written from, in the order journal.py
// declares its fields.
//
// The list is spelled out rather than derived from the struct tags because it
// is the contract with the Python writer, not a restatement of this file: a
// field added here without a writer that writes it should read as a
// disagreement, which a derived list would hide.
var entryKeys = []string{
	"node", "kind", "input_hash", "delta", "outcome",
	"tools", "effort", "tokens", "seconds", "detail",
}

// decode turns one line into an Entry, or says why it is not one.
//
// The object is read twice on purpose: the first pass sees which keys are
// *present*, which unmarshalling into a struct cannot tell apart from absent.
// The difference matters because `tools`, `effort` and `detail` are `str | None`
// on the Python side -- a `null` there is a written value and its key is always
// there, so absence is damage and `null` is not.
func decode(line string) (Entry, error) {
	var keyed map[string]json.RawMessage
	if err := json.Unmarshal([]byte(line), &keyed); err != nil {
		return Entry{}, err
	}

	var missing []string
	for _, key := range entryKeys {
		if _, ok := keyed[key]; !ok {
			missing = append(missing, key)
		}
	}
	if len(missing) > 0 {
		return Entry{}, fmt.Errorf("no value for %s", strings.Join(missing, ", "))
	}

	var unknown []string
	for key := range keyed {
		if !slices.Contains(entryKeys, key) {
			unknown = append(unknown, key)
		}
	}
	if len(unknown) > 0 {
		// Sorted, because a map's order is not stable and an error message that
		// changed between runs would be one no test could name.
		slices.Sort(unknown)
		return Entry{}, fmt.Errorf("unknown key %s", strings.Join(unknown, ", "))
	}

	var entry Entry
	if err := json.Unmarshal([]byte(line), &entry); err != nil {
		return Entry{}, err
	}
	return entry, nil
}
